### ISSUE TITLE
Please create an official upspinserver Docker container

### ISSUE THREAD

**Issue by stapelberg** [upspin/upspin#239]
I’d like to run upspinserver via Docker. Could you create an official upspinserver Docker container please? If you need help, let me know and I could take a stab at it.

**Comment by adg** [comment]
How do official Docker containers usually deal with persistent state? The upspinserver writes directory server logs and (optionally) block data to local disk.

Incidentally, in what context would you use the Docker image? The reason our [setup instructions](https://upspin.io/doc/server_setup.md) don't use is that I couldn't see how it would make deployment easier.

**Comment by stapelberg** [comment]
> How do official Docker containers usually deal with persistent state? The upspinserver writes directory server logs and (optionally) block data to local disk.

Usually, containers define a volume and users take care of mounting it appropriately. An example is the [official postgres container](https://hub.docker.com/r/library/postgres/), which defines a volume `/var/lib/postgresql/data`.

> Incidentally, in what context would you use the Docker image? The reason our setup instructions don't use is that I couldn't see how it would make deployment easier.

Having a docker image would allow me to easily run upspin on Google Container Engine (GKE), a custom-hosted kubernetes instance or a more bare-bones setup like a CoreOS based machine or cluster, where one can easily start docker containers after configuring them in cloud-init.

**Comment by DeedleFake** [comment]
I don't know about an official one, but I threw [one](https://hub.docker.com/r/deedlefake/upspinserver) together really quickly. It's a from scratch image, so it's got nothing in it other than the binary.

**Comment by adg** [comment]
> Having a docker image would allow me to easily run upspin on Google Container Engine (GKE), a custom-hosted kubernetes instance or a more bare-bones setup like a CoreOS based machine or cluster, where one can easily start docker containers after configuring them in cloud-init.

We already have Dockerfiles and a deployment system for deploying Upspin to GKE: https://upspin.io/doc/deploying_to_google_cloud.md

There are Dockerfiles for storeserver and dirserver in the repo inside cloud/docker. But this isn't the deployment approach we decided to endorse, hence why the doc I just mentioned isn't linked from the home page, as there are so many moving parts and it's not clear that GKE+Docker buys us anything.

**Comment by swetland** [comment]
I think just a recipe for assembling a scratch docker container would be nice.  I suspect it might be smoother than mucking about with systemd (ew), xinited (ew), creating users, and whatnot (which I did, adapting the Ubuntu setup instructions for Debian).

**Comment by awly** [comment]
> I suspect it might be smoother than mucking about with systemd (ew), xinited (ew), creating users, and whatnot (which I did, adapting the Ubuntu setup instructions for Debian).

As a side note, i'm not sure why Ubuntu setup instructions insist on using port 443 (and having to deal with xinited). It seems that upspin should work just fine with arbitrary ports.

**Comment by adg** [comment]
@awly IIUC LetsEncrypt requires that you be listening on port 443 to issue certs.

**Comment by adg** [comment]
@swetland once you have the Docker container how do you run it? I agree systemd and xinetd are icky, but what's the replacement for those things?

**Comment by stapelberg** [comment]
> IIUC LetsEncrypt requires that you be listening on port 443 to issue certs.

Yes, this is a current limitation of the default challenge mechanism which LetsEncrypt uses. https://github.com/letsencrypt/acme-spec/issues/33 is related, but it’s not clear to me where one can easily see the current state of things.

I think we should document that the requirement for using port 443 comes from LetsEncrypt so that users don’t need to wonder, and can make an informed choice to use a different port should they not require upspinserver’s built-in LetsEncrypt functionality.

> once you have the Docker container how do you run it? I agree systemd and xinetd are icky, but what's the replacement for those things?

You’d run it the same way you run any other container: using fleet, kubernets, swarm, systemd, $favorite_tool_here. Why are you asking? :)

**Comment by adg** [comment]
> You’d run it the same way you run any other container: using fleet, kubernets, swarm, systemd, $favorite_tool_here. Why are you asking? :)

So how is that any less onerous than just running the upspinserver binary under systemd? 😕

**Comment by stapelberg** [comment]
The most significant difference for me is who’s responsible for updates: CoreOS (and others) deploy updates automatically. As a side-effect of an update-triggered reboot (happening every few weeks), all Docker containers I run get pulled automatically.

When running a binary under systemd, I need to keep that binary up to date. Currently, that means I need to manually run `go get -u`, `go build` or `go install` and `scp`.

If upspin was already packaged for Debian, I could install the upspin package on my server and use unattended-upgrades. However, I think it’ll be a while before the stable versions of commonly used distributions include upspin, given their long release cycles.

In the interim, and for users of Docker-centric distributions, I think it’d be good to have an official Docker container.

**Comment by DeedleFake** [comment]
> So how is that any less onerous than just running the upspinserver binary under systemd?

It depends on the use case. The Docker image that I linked above I'm using to run upspinserver behind Traefik. Traefik can coordinate with the Docker daemon to get configuration for each service from the container's definition, which makes things very clean. It also means that I can keep the service from being reached by anything except through the containers exposed interface, which gives quite a bit of fine control over it. I can also fiddle with things like allowed memory usage pretty easily.

**Comment by awly** [comment]
@adg what do you think about suggesting setcap as an alternative to xinited in the docs?
This is subjective but i prefer to run as few services as possible.

**Comment by adg** [comment]
@awly yes, please! I used `xinetd` because that was what I knew. If there's a simpler way, let's do it. Would you mind sending the doc change?

**Comment by awly** [comment]
@adg sent https://upspin-review.googlesource.com/c/8031/

**Comment by stapelberg** [comment]
Here’s one more use-case: docker containers would allow to run upspinserver easily on storage appliances such as FreeNAS, see http://www.freenas.org/blog/introducing-freenas-corral-worlds-first-open-source-hyper-converged-storage-platform/

**Comment by adg** [comment]
We should definitely do this. It would let us automate the creation of the Google Cloud Compute (or AWS EC2) instance that runs upspinserver, and therefore automate most of the setup process for users who want to use a cloud provider.

**Comment by noonien** [comment]
The main benefit of having a docker image is the ability to run upspinserver in kubernetes. It would be great seeing a helm package as well.

**Comment by adg** [comment]
I've made an image available publicly as `gcr.io/upspin-containers/upspinserver`, and is prepared by [upspin-deploy-gcp](https://godoc.org/gcp.upspin.io/cmd/upspin-deploy-gcp/).

I'm now working on automatically updating that image.

**Comment by adg** [comment]
The image is now automatically updated on each push to the gcp repo. All that's left to close off this issue is to document the existence of the docker image.
